internal/analysis/rundynamic.go

```diff
@@ -0,0 +1,111 @@
+			}
+			return results, err
+		}
+
+		results[phase] = result
+		lastNonErrorPhase = phase
+
+		if result == nil || result.Status != dynamicanalysis.StatusCompleted {
+			// Error caused by an issue with the package (probably).
+			// Don't continue processing subsequent phases if this one did not complete successfully.
+			break
+		}
+	}
+	if logStatus {
+		// Produce a log message for the final status to help generate metrics.
+		logDynamicAnalysisResult(pkg, lastNonErrorPhase, results[lastNonErrorPhase])
+	}
+
+	return results, nil
+}
+
+// logDynamicAnalysisError
+// Indicates some error happened while attempting to run the package code, which was not caused by the package itself.
+// This kind of error means it was not possible to analyse the package properly, and the results are invalid.
+func logDynamicAnalysisError(pkg *pkgecosystem.Pkg, errorPhase pkgecosystem.RunPhase, err error) {
+	log.Error("Analysis run failed",
+		log.Label("ecosystem", pkg.Ecosystem()),
+		log.Label("name", pkg.Name()),
+		log.Label("phase", string(errorPhase)),
+		log.Label("version", pkg.Version()),
+		"error", err)
+}
+
+// logDynamicAnalysisResult
+// Indicates that the package code was run successfully, and what happened when it was run
+// This may include errors in the analysis of the package, but not errors in the running itself.
+func logDynamicAnalysisResult(pkg *pkgecosystem.Pkg, finalPhase pkgecosystem.RunPhase, finalResult *dynamicanalysis.Result) {
+	ecosystem := pkg.Ecosystem()
+	name := pkg.Name()
+	version := pkg.Version()
+	lastPhase := string(finalPhase)
+
+	if finalResult == nil {
+		log.Warn("Analysis error - no result for last phase",
+			log.Label("ecosystem", ecosystem),
+			log.Label("name", name),
+			log.Label("version", version),
+			log.Label("last_phase", lastPhase))
+		return
+	}
+
+	switch finalResult.Status {
+	case dynamicanalysis.StatusCompleted:
+		log.Info("Analysis completed sucessfully",
+			log.Label("ecosystem", ecosystem),
+			log.Label("name", name),
+			log.Label("version", version),
+			log.Label("last_phase", lastPhase))
+
+	case dynamicanalysis.StatusErrorAnalysis:
+		log.Warn("Analysis error - analysis",
+			log.Label("ecosystem", ecosystem),
+			log.Label("name", name),
+			log.Label("version", version),
+			log.Label("last_phase", lastPhase))
+	case dynamicanalysis.StatusErrorTimeout:
+		log.Warn("Analysis error - timeout",
+			log.Label("ecosystem", ecosystem),
+			log.Label("name", name),
+			log.Label("version", version),
+			log.Label("last_phase", lastPhase))
+	case dynamicanalysis.StatusErrorOther:
+		log.Warn("Analysis error - other",
+			log.Label("ecosystem", ecosystem),
+			log.Label("name", name),
+			log.Label("version", version),
+			log.Label("last_phase", lastPhase))
+	}
+}
+
+// phasesToAnalyze
+// Logic to decide which phases of dynamic analysis to run
+// Currently just returns all possible phases for the package ecosystem
+func phasesToAnalyze(pkg *pkgecosystem.Pkg) []pkgecosystem.RunPhase {
```

Comment:
I think this function can be removed and the call below made inline.

If we need to wrap it in a function in the future it is super easy to do. We don't need to incur the extra readability cost now.

Reply:
Okay